### PR TITLE
MLR-245: Implement length checks on site number depending on site type.

### PR DESCRIPTION
**Pull request by kmschoep-usgs**

### CODE REVIEW

File: mlrvalidator/validators/reference.py [hunk 1]
```diff
@@ -104,3 +104,16 @@ def get_land_net_templates(self, district_code):
...
         return land_net_template
+
+
+class SiteNumberFormat(ReferenceInfo):
+    def get_site_number_template(self, site_type_code):
+        site_number_format_refs = self.reference_info['siteNumberFormatCodes']
+
+        try:
+            idx = next(index for (index, d) in enumerate(site_number_format_refs) if site_type_code in d['siteTypeCode'])
+            site_number_format_ref = site_number_format_refs[idx]['siteNumberFormatCode']
```
Comment (ayan-usgs) [inline mlrvalidator/validators/reference.py]: I'd recommend moving this to an `else` clause to keep the scope of the `try` narrow.

File: mlrvalidator/validators/cross_field_ref_error_validator.py [hunk 2]
```diff
@@ -164,6 +165,33 @@ def _validate_state_longitude_range(self):
                     if not (state_attr['state_min_long_va'] <= lat < state_attr['state_max_long_va']):
                         self._errors['longitude'] = ['Longitude is out of range for state {0}'.format(state)]
 
+    def _validate_site_number_format(self):
+        '''
+        :return: boolean
+        '''
+        keys = ['siteNumber', 'siteTypeCode']
+        if self._any_fields_in_document(keys):
+            site_number, site_type_code = [self.merged_document.get(key, '').strip() for key in keys]
```
Comment (ayan-usgs) [inline mlrvalidator/validators/cross_field_ref_error_validator.py]: I'm not sure how much mileage we're getting with this list comprehension. It seems like it'd be simpler and more readable to call `self.merged_document.get()` twice.

**Comment by ayan-usgs** [comment]
Looks good to go. I don't think there are any major show stoppers.

File: mlrvalidator/validators/cross_field_ref_error_validator.py (hunk 2 above)
Comment (mbucknell) [reply mlrvalidator/validators/cross_field_ref_error_validator.py]: It follows a pattern that we use in many places.

File: mlrvalidator/references/site_number_format.json [hunk 3]
```diff
@@ -0,0 +1,77 @@
+{
```
Comment (mbucknell) [inline mlrvalidator/references/site_number_format.json]: Just verifying that this information is not pulled from a query. If it is, we should add the code to the json_population_scripts directory.

File: mlrvalidator/validators/cross_field_ref_error_validator.py (hunk 2 above)
Comment (ayan-usgs) [reply mlrvalidator/validators/cross_field_ref_error_validator.py]: Okay.

File: mlrvalidator/validators/tests/test_cross_field_ref_error_validator.py [hunk 4]
```diff
@@ -29,8 +30,9 @@ class CrossFieldRefValidatorForCountiesTestCase(TestCase):
     @mock.patch('mlrvalidator.validators.cross_field_ref_error_validator.NationalWaterUseCodes')
     @mock.patch('mlrvalidator.validators.cross_field_ref_error_validator.SiteTypesCrossField')
     @mock.patch('mlrvalidator.validators.cross_field_ref_error_validator.LandNetCrossField')
+    @mock.patch('mlrvalidator.validators.cross_field_ref_error_validator.SiteNumberFormat')
     @mock.patch('mlrvalidator.validators.cross_field_ref_error_validator.States')
-    def setUp(self, mstates_ref, msite_type_ref, mland_net_ref, mwater_use_ref, mref_validator_class):
+    def setUp(self, mstates_ref, msite_type_ref, mland_net_ref, msite_number_ref, mwater_use_ref, mref_validator_class):
```
Comment (mbucknell) [inline mlrvalidator/validators/tests/test_cross_field_ref_error_validator.py]: We should probably keep the pattern of mocking SiteNumberFormat in all of the tests except where we are specifically testing it.

File: mlrvalidator/references/site_number_format.json (hunk 3 above)
Comment (kmschoep-usgs) [reply mlrvalidator/references/site_number_format.json]: It's hand-built from the GW coding document.

**Pull request merged**